types/lodash/common/lang.d.ts

```diff
@@ -569,6 +571,12 @@ declare module "../index" {
          * @param value The value to inspect.
          * @return Returns true if value is empty, else false.
          */
+        isEmpty<T extends { __trapAny: any }>(value?: T): boolean;
+        isEmpty(value: string | null | undefined): value is '' | null | undefined;
+        isEmpty(value: any[] | null | undefined): boolean;
+        isEmpty(value: ReadonlyArray<any> | null | undefined): value is Readonly<[]> | null | undefined;
```

Comment:
This overload scares me the most - I'm not sure if TypeScript treats `Readonly<[]>` differently from `ReadonlyArray` in a way could breaks people's existing code. But I don't have any concrete examples so maybe this is fine (no action required).

Reply:
Yeah, i was doing some experimentation, and it fails in the same way this https://github.com/DefinitelyTyped/DefinitelyTyped/pull/60401#issuecomment-1130231313 does. It seems to be a problem with type guards - either they refine a type (like `string -> ''`), or they split up a union (like `string | undefined` -> `undefined`), but won't do `string | undefined -> '' | undefined`, the type just changes to `undefined`. I couldn't find a bug in typescript for this, but I didnt spend too much time digging.

So we might need to revert / cancel, or explain how to opt-out of type guards using module augmentation, or add an opt-in. Unfortunately, there's not a perfect solution here, which sucks.

Reply:
candidates for typescript bugs:
- https://github.com/microsoft/TypeScript/issues/43904
- https://github.com/microsoft/TypeScript/issues/44382

I'll investigate with older versions of typescript on the weekend.

Reply:
I try to test it, and turns out:
```typescript
const valA = '';

if (isEmpty(valA)) {
  valA; // valA type is ""
} else {
  valA; // valA type is never. which we want is string.
}

// but if i change to:
const valA : string = '';

if (isEmpty(valA)) {
  valA; // valA type is ""
} else {
  valA; // valA type is string.
}
```
so `valA` must be explicitly typed string otherwise the type guards will fail.

Reply:
@forivall That's unfortunate, they may net be a good solution. We may need to return `boolean` for arrays to be on the safe side.

@fatihaziz that sounds like intended behavior to me. Without the explicit type, TS will infer the type to be `''`.

Reply:
Ohh right,, so it doesn't matter